analyse error: Unable to get artifact

The following error is emitted when building the example project at https://github.com/kwin/aem-analyser-example/tree/analyse-failed-with-wrong-maven-gav

```
[INFO] --- aemanalyser-maven-plugin:0.0.14:aggregate (default-aggregate) @ aem-analyser-example ---
[INFO] Using SDK Version for analysis: Dependency {groupId=com.adobe.aem, artifactId=aem-sdk-api, version=2020.11.4506.20201112T235200Z-201028, type=jar}
[INFO]
[INFO] --- aemanalyser-maven-plugin:0.0.14:analyse (default-analyse) @ aem-analyser-example ---
[INFO] Starting analyzing feature 'biz.netcentric.aem:aem-analyser-example:slingosgifeature:aggregated:0.0.1-SNAPSHOT'...
Downloading from nexus: https://repo.int.netcentric.biz/nexus/content/groups/public/bcpkix/bcpkix/1.64/bcpkix-1.64.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  9.957 s
[INFO] Finished at: 2020-11-20T09:15:03+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.adobe.aem:aemanalyser-maven-plugin:0.0.14:analyse (default-analyse) on project aem-analyser-example: Exception during analysing feature biz.netcentric.aem:aem-analyser-example:slingosgifeature:aggregated:0.0.1-SNAPSHOT : Unable to get artifact for bcpkix:bcpkix:1.64: Could not find artifact bcpkix:bcpkix:jar:1.64 in nexus (https://repo.int.netcentric.biz/nexus/content/groups/public/)
[ERROR]
[ERROR] Try downloading the file manually from the project website.
[ERROR]
[ERROR] Then, install it using the command:
[ERROR]     mvn install:install-file -DgroupId=bcpkix -DartifactId=bcpkix -Dversion=1.64 -Dpackaging=jar -Dfile=/path/to/file
[ERROR]
[ERROR] Alternatively, if you host your own repository you can deploy the file there:
[ERROR]     mvn deploy:deploy-file -DgroupId=bcpkix -DartifactId=bcpkix -Dversion=1.64 -Dpackaging=jar -Dfile=/path/to/file -Durl=[url] -DrepositoryId=[id]
[ERROR]
[ERROR]
[ERROR]   bcpkix:bcpkix:jar:1.64
[ERROR]
[ERROR] from the specified remote repositories:
[ERROR]   nexus (https://repo.int.netcentric.biz/nexus/content/groups/public/, releases=true, snapshots=true)
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

Obviously it uses the wrong GAV for artifact https://search.maven.org/artifact/org.bouncycastle/bcpkix-jdk15on/1.64/jar which is referenced transitively in https://github.com/Netcentric/accesscontroltool/blob/c82f6adc78ceeee0b511001e43b81379dc13d3da/accesscontroltool-package/pom.xml#L110. Obviously something is wrong while determining the GAV of this bundle.

When I build the test project I get the following error before it hits your error:

    [ERROR] Failed to execute goal com.adobe.aem:aemanalyser-maven-plugin:0.0.14:convert (default-convert) on project aem-analyser-example: Content Package Converter Exception: org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact biz.netcentric.cq.tools.accesscontroltool:accesscontroltool-package:zip:2.6.2 -> [Help 1]

I guess the test case depends on this content package that is not available publicly...

--- a/aemanalyser-maven-plugin/pom.xml
+++ b/aemanalyser-maven-plugin/pom.xml
@@ -112,7 +112,7 @@ governing permissions and limitations under the License.
     <dependency>
       <groupId>org.apache.sling</groupId>
       <artifactId>slingfeature-maven-plugin</artifactId>
-      <version>1.4.12</version>
+      <version>1.4.17-SNAPSHOT</version>
     </dependency>
     
     <!-- Test -->

--- a/aemanalyser-maven-plugin/src/main/java/com/adobe/aem/analyser/mojos/AnalyseMojo.java
+++ b/aemanalyser-maven-plugin/src/main/java/com/adobe/aem/analyser/mojos/AnalyseMojo.java
@@ -17,9 +17,15 @@ import org.apache.maven.plugin.MojoFailureException;
 import org.apache.maven.plugins.annotations.LifecyclePhase;
 import org.apache.maven.plugins.annotations.Mojo;
 import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.sling.feature.ArtifactId;
+import org.apache.sling.feature.builder.ArtifactProvider;
+import org.apache.sling.feature.io.artifacts.ArtifactManager;
+import org.apache.sling.feature.io.artifacts.ArtifactManagerConfig;
 import org.apache.sling.feature.maven.mojos.AnalyseFeaturesMojo;
 import org.apache.sling.feature.maven.mojos.Scan;
 
+import java.io.IOException;
+import java.net.URL;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
@@ -32,6 +38,8 @@ import static com.adobe.aem.analyser.mojos.MojoUtils.setParameter;
 @Mojo(name = "analyse", defaultPhase = LifecyclePhase.TEST)
 public class AnalyseMojo extends AnalyseFeaturesMojo {
     boolean unitTestMode = false;
+
+    ArtifactManager localArtifactManager;
 
     @Parameter(defaultValue =
         "requirements-capabilities,"
@@ -49,7 +57,33 @@ public class AnalyseMojo extends AnalyseFeaturesMojo {
     Map<String, Properties> taskConfiguration;
 
     @Override
+    protected ArtifactProvider getArtifactProvider() {
+        ArtifactProvider ap = super.getArtifactProvider();
+
+        return new ArtifactProvider() {
+            @Override
+            public URL provide(ArtifactId id) {
+                if (localArtifactManager != null) {
+                    URL url = localArtifactManager.provide(id);
+                    if (url != null)
+                        return url;
+                }
+
+                return ap.provide(id);
+            }
+        };
+    }
+
+    @Override
     public void execute() throws MojoExecutionException, MojoFailureException {
+        try {
+            ArtifactManagerConfig amcfg = new ArtifactManagerConfig();
+            amcfg.setRepositoryUrls(new String[] { MojoUtils.getConversionOutputDir(project).toURI().toURL().toString() });
+            localArtifactManager = ArtifactManager.getArtifactManager(amcfg);
+        } catch (IOException e) {
+            throw new MojoExecutionException("Problem configuring Artifact Provider for :" + MojoUtils.getConversionOutputDir(project));
+        }
+
         if (taskConfiguration == null) {
             taskConfiguration = new HashMap<>();
         }

--- a/aemanalyser-maven-plugin/src/test/java/com/adobe/aem/analyser/mojos/AnalyseMojoTest.java
+++ b/aemanalyser-maven-plugin/src/test/java/com/adobe/aem/analyser/mojos/AnalyseMojoTest.java
@@ -12,13 +12,28 @@
 
 package com.adobe.aem.analyser.mojos;
 
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.model.Build;
 import org.apache.maven.project.MavenProject;
+import org.apache.sling.feature.ArtifactId;
+import org.apache.sling.feature.builder.ArtifactProvider;
+import org.apache.sling.feature.io.artifacts.ArtifactManager;
+import org.apache.sling.feature.io.artifacts.ArtifactManagerConfig;
 import org.apache.sling.feature.maven.mojos.Scan;
+import org.junit.After;
+import org.junit.Before;
 import org.junit.Test;
 import org.mockito.Mockito;
 
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
@@ -29,9 +44,30 @@ import java.util.Set;
 import static org.junit.Assert.assertEquals;
 
 public class AnalyseMojoTest {
+
+    private Path tempDir;
+
+    @Before
+    public void setUp() throws IOException {
+        tempDir = Files.createTempDirectory(getClass().getSimpleName());
+    }
+
+    @After
+    public void tearDown() throws IOException {
+        // Delete the temp dir
+        Files.walk(tempDir)
+            .sorted(Comparator.reverseOrder())
+            .map(Path::toFile)
+            .forEach(File::delete);
+    }
+
     @Test
     public void testExecute() throws Exception {
+        Build build = Mockito.mock(Build.class);
+        Mockito.when(build.getDirectory()).thenReturn(tempDir.toString());
+
         MavenProject prj = Mockito.mock(MavenProject.class);
+        Mockito.when(prj.getBuild()).thenReturn(build);
         Mockito.when(prj.getContextValue(AggregateWithSDKMojo.class.getName() + "-aggregates"))
             .thenReturn(new HashSet<>(Arrays.asList("agg1", "agg2")));
 
@@ -67,6 +103,7 @@ public class AnalyseMojoTest {
     @Test
     public void testAddTaskConfig() throws Exception {
         MavenProject prj = Mockito.mock(MavenProject.class);
+        Mockito.when(prj.getBuild()).thenReturn(Mockito.mock(Build.class));
         Mockito.when(prj.getContextValue(AggregateWithSDKMojo.class.getName() + "-aggregates"))
             .thenReturn(Collections.singleton("aggregates"));
 
@@ -100,6 +137,106 @@ public class AnalyseMojoTest {
         assertEquals("Default task configuration not as expected",
                 "global,com.adobe.aem.deprecated,com.adobe.aem.internal",
                 scan.getTaskConfiguration().get("api-regions-check-order").get("order"));
+    }
+
+    @Test
+    public void testLocalArtifactManager() throws Exception {
+        File ab123 = Files.createTempFile(tempDir, "test", ".tmp").toFile();
+
+        Path dir = tempDir.resolve("cp-conversion/d/e/f/456");
+        Files.createDirectories(dir);
+        Path def456 = dir.resolve("f-456.jar");
+        Files.createFile(def456);
+
+        Build build = Mockito.mock(Build.class);
+        Mockito.when(build.getDirectory()).thenReturn(tempDir.toString());
+
+        List<Artifact> attached = new ArrayList<>();
+        Artifact a = Mockito.mock(Artifact.class);
+        Mockito.when(a.getGroupId()).thenReturn("a");
+        Mockito.when(a.getArtifactId()).thenReturn("b");
+        Mockito.when(a.getVersion()).thenReturn("123");
+        Mockito.when(a.getType()).thenReturn("jar");
+        Mockito.when(a.getFile()).thenReturn(ab123);
+        attached.add(a);
+
+        MavenProject prj = Mockito.mock(MavenProject.class);
+        Mockito.when(prj.getAttachedArtifacts()).thenReturn(attached);
+        Mockito.when(prj.getBuild()).thenReturn(build);
+        Mockito.when(prj.getContextValue(AggregateWithSDKMojo.class.getName() + "-aggregates"))
+            .thenReturn(Collections.singleton("aggregates"));
+
+        AnalyseMojo mojo = new TestAnalyseMojo(prj);
+        mojo.unitTestMode = true;
+        mojo.includeTasks = Collections.singletonList("mytask");
+
+        mojo.execute();
+
+        ArtifactProvider ap = mojo.getArtifactProvider();
+        URL url = ap.provide(ArtifactId.fromMvnId("a:b:123"));
+        assertEquals(ab123.toURI().toURL(), url);
+
+        URL url2 = ap.provide(ArtifactId.fromMvnId("d.e:f:456"));
+        assertEquals(def456.toUri().toURL(), url2);
+    }
+
+    @Test
+    public void testCustomArtifactProvider() throws Exception {
+        File ab123 = Files.createTempFile(tempDir, "test", ".tmp").toFile();
+
+        Path dir = tempDir.resolve("d/e/f/456");
+        Files.createDirectories(dir);
+        Path def456 = dir.resolve("f-456.jar");
+        Files.createFile(def456);
+
+        List<Artifact> attached = new ArrayList<>();
+        Artifact a = Mockito.mock(Artifact.class);
+        Mockito.when(a.getGroupId()).thenReturn("a");
+        Mockito.when(a.getArtifactId()).thenReturn("b");
+        Mockito.when(a.getVersion()).thenReturn("123");
+        Mockito.when(a.getType()).thenReturn("jar");
+        Mockito.when(a.getFile()).thenReturn(ab123);
+        attached.add(a);
+
+        MavenProject prj = Mockito.mock(MavenProject.class);
+        Mockito.when(prj.getAttachedArtifacts()).thenReturn(attached);
+
+        ArtifactManagerConfig amc = new ArtifactManagerConfig();
+        amc.setRepositoryUrls(new String [] { tempDir.toUri().toString() });
+        ArtifactManager lam = ArtifactManager.getArtifactManager(amc);
+
+        AnalyseMojo mojo = new TestAnalyseMojo(prj);
+        mojo.localArtifactManager = lam;
+
+        ArtifactProvider ap = mojo.getArtifactProvider();
+        URL url = ap.provide(ArtifactId.fromMvnId("a:b:123"));
+        assertEquals(ab123.toURI().toURL(), url);
+
+        URL url2 = ap.provide(ArtifactId.fromMvnId("d.e:f:456"));
+        assertEquals(def456.toUri().toURL(), url2);
+    }
+
+    @Test
+    public void testNoCustomArtifactProvider() throws Exception {
+        File ab123 = Files.createTempFile(tempDir, "test", ".tmp").toFile();
+
+        List<Artifact> attached = new ArrayList<>();
+        Artifact a = Mockito.mock(Artifact.class);
+        Mockito.when(a.getGroupId()).thenReturn("a");
+        Mockito.when(a.getArtifactId()).thenReturn("b");
+        Mockito.when(a.getVersion()).thenReturn("123");
+        Mockito.when(a.getType()).thenReturn("slingosgifeature");
+        Mockito.when(a.getFile()).thenReturn(ab123);
+        attached.add(a);
+
+        MavenProject prj = Mockito.mock(MavenProject.class);
+        Mockito.when(prj.getAttachedArtifacts()).thenReturn(attached);
+
+        AnalyseMojo mojo = new TestAnalyseMojo(prj);
+        ArtifactProvider ap = mojo.getArtifactProvider();
+        URL url = ap.provide(ArtifactId.fromMvnId("a:b:slingosgifeature:123"));
+        assertEquals(ab123.toURI().toURL(), url);
+
     }
 
     private Set<String> getSelectionInstructions(List<?> sels, String type) throws Exception {

